### PR TITLE
feat: TrafficRouting support with AWS App Mesh (#1401)

### PR DESCRIPTION
**Pull request by kiranmeduri**
TrafficRouting support with AWS App Mesh

- Supports canary based strategy with weighted traffic splitting using App Mesh
- Added getting-started doc for App Mesh

https://github.com/argoproj/argo-rollouts/issues/1401

Organization: Amazon Web Services

Signed-off-by: Kiran Meduri <meduri@amazon.com>

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-rollouts/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I've signed my commits with [DCO](https://github.com/argoproj/argoproj)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My builds are green. Try syncing with master if they are not. 
* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo-rollouts/blob/master/USERS.md).

### CODE REVIEW

**Comment by jessesuen** [comment]
@kiranmeduri sorry for not looking at this sooner. I'd like to start reviewing this for potential inclusion in the upcoming 1.2 release. Is this ready for review? Some of the tests/linting are failing

**Comment by kiranmeduri** [comment]
@jessesuen thanks for update, let me sync again and check the failures. I will update the PR and notify.

**Comment by kiranmeduri** [comment]
@jessesuen PR is ready for review but unclear how to fix my "verify codegen" check that is failing. Can you point me in right direction? Thanks

**Comment by jessesuen** [comment]
> @jessesuen PR is ready for review but unclear how to fix my "verify codegen" check that is failing. Can you point me in right direction? Thanks

Did you try running `make codegen` target? That should fix it.

**Comment by kiranmeduri** [comment]
> > @jessesuen PR is ready for review but unclear how to fix my "verify codegen" check that is failing. Can you point me in right direction? Thanks
> 
> Did you try running `make codegen` target? That should fix it.

Yes i did and `make codegen` succeeds. However, I notice that it generates new versions of manifest file as my controller-gen is 0.5.0 instead of upstream 0.7.0. Do you think that can be the culprit? In the meantime I will update that and see.

Thanks

**Comment by kiranmeduri** [comment]
@jessesuen Build steps is failing on github due to `TestPauseRolloutAfterInconclusiveExperiment` failure. I notice this test to be a bit flaky with the following error.

```
FAIL: TestPauseRolloutAfterInconclusiveExperiment (0.20s)
    experiment_test.go:364:
                Error Trace:    experiment_test.go:364
                Error:          Not equal:
                                expected: "{\"status\":{\"canary\":{\"currentExperiment\":null},\"conditions\":[{\"lastTransitionTime\":\"2022-02-01T01:16:42Z\",\"lastUpdateTime\":\"2022-02-01T01:16:42Z\",\"message\":\"Rollout \\\"foo\\\" is progressing
.\",\"reason\":\"ReplicaSetUpdated\",\"status\":\"True\",\"type\":\"Progressing\"},{\"lastTransitionTime\":\"2022-02-01T01:16:42Z\",\"lastUpdateTime\":\"2022-02-01T01:16:42Z\",\"message\":\"Rollout has minimum availability\",\"reason\":\"AvailableReason
\",\"status\":\"True\",\"type\":\"Available\"}],\"controllerPause\":true,\"message\":\"InconclusiveExperiment\",\"observedGeneration\":\"124\",\"pauseConditions\":[{\"reason\":\"InconclusiveExperiment\",\"startTime\":\"2022-02-01T01:16:43Z\"}],\"phase\"
:\"Paused\"}}"
                                actual  : "{\"status\":{\"canary\":{\"currentExperiment\":null},\"conditions\":[{\"lastTransitionTime\":\"2022-02-01T01:16:42Z\",\"lastUpdateTime\":\"2022-02-01T01:16:42Z\",\"message\":\"Rollout \\\"foo\\\" is progressing
.\",\"reason\":\"ReplicaSetUpdated\",\"status\":\"True\",\"type\":\"Progressing\"},{\"lastTransitionTime\":\"2022-02-01T01:16:42Z\",\"lastUpdateTime\":\"2022-02-01T01:16:42Z\",\"message\":\"Rollout has minimum availability\",\"reason\":\"AvailableReason
\",\"status\":\"True\",\"type\":\"Available\"}],\"controllerPause\":true,\"message\":\"InconclusiveExperiment\",\"observedGeneration\":\"124\",\"pauseConditions\":[{\"reason\":\"InconclusiveExperiment\",\"startTime\":\"2022-02-01T01:16:42Z\"}],\"phase\"
:\"Paused\"}}"
```
I can see that the expectation is failing on `startTime` diff. Can you confirm if this test is flaky or if I have something missing?

**Comment by kiranmeduri** [comment]
@jessesuen this PR is ready for review. Please take a look and I would really appreciate if we can get it into next release (1.2). Thanks

File: pkg/apis/rollouts/validation/validation_references.go [hunk 1]
```diff
@@ -321,6 +326,57 @@ func ValidateAmbassadorMapping(obj unstructured.Unstructured) field.ErrorList {
...
+		return field.ErrorList{err}
+	}
+	return field.ErrorList{}
+}
+
+func ValidateAppMeshVirtualRouter(vrouter *unstructured.Unstructured) *field.Error {
+	routesFldPath := field.NewPath("spec", "routes")
+	allRoutesI, found, err := unstructured.NestedSlice(vrouter.Object, "spec", "routes")
+	if !found || err != nil || len(allRoutesI) == 0 {
+		msg := fmt.Sprintf("No routes for defined for AppMesh virtual-router %s", vrouter.GetName())
```
Comment (harikrongali) [inline pkg/apis/rollouts/validation/validation_references.go:347]: nit: "No routes defined for AppMesh virtual-router %s"

File: rollout/trafficrouting/appmesh/appmesh.go [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+			return false, field.Invalid(wtFldPath, uVr.GetName(), ErrNotWellFormed)
+		}
+		wtVnRefFldPath := wtFldPath.Child("virtualNodeRef")
+		wtVnRef, ok := wt["virtualNodeRef"].(map[string]interface{})
+		if !ok {
+			return false, field.Invalid(wtVnRefFldPath, uVr.GetName(), ErrNotWellFormed)
+		}
+		wtVnName, _ := wtVnRef["name"].(string)
+		wtVnNamespace := defaultIfEmpty(wtVnRef["namespace"], r.rollout.Namespace)
+		weight, err := toInt64(wt["weight"])
```
Comment (harikrongali) [inline rollout/trafficrouting/appmesh/appmesh.go:241]: why do you need this be Int64?

File: rollout/trafficrouting/appmesh/appmesh.go [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+		return make(map[string]interface{}), nil
+	}
+	return m, nil
+}
+
+func setPodSelectorMatchLabels(vnode *unstructured.Unstructured, ml map[string]interface{}) error {
+	return unstructured.SetNestedMap(vnode.Object, ml, "spec", "podSelector", "matchLabels")
+}
+
+func toInt64(obj interface{}) (int64, error) {
```
Comment (harikrongali) [inline rollout/trafficrouting/appmesh/appmesh.go:335]: seems not required as int32 is good to us like other implementations

**Comment by harikrongali** [comment]
@kiranmeduri few nit picks other than that lgtm. 
can you please rebase to master

File: rollout/trafficrouting/appmesh/appmesh.go (hunk 2 above)
Comment (kiranmeduri) [reply rollout/trafficrouting/appmesh/appmesh.go:241]: AppMesh CRD uses int64 in [here](https://aws.github.io/aws-app-mesh-controller-for-k8s/reference/api_spec/#appmesh.k8s.aws/v1beta2.WeightedTarget), hence needed to do this.

**Review by harikrongali — APPROVED** [review]

**Review by jessesuen — APPROVED** [review]
Great work!

**Pull request merged**